Post: `build.commands` (#195)
readthedocs/blog: build-customization.rst

.. post:: Nov 10, 2022
   :tags: announcement, feature, builds
   :author: Manuel
   :location: BCN

.. meta::
   :description lang=en:

>       We released user-defined commands (``build.commands`` config key)
>       which allow users to overwrite the build process completely.

Reviewer: ```suggestion
      We released a user-defined command (``build.commands`` config key)
      which allow you to overwrite the build process completely.
```

Author: Hrm, it's not "a user-defined command", but I think it not correct what I wrote either.

Probably,

```suggestion
      We released a new config key (``build.commands``) to specify user-defined commands
      which allow users to overwrite the build process completely.
```

is that better?

Reviewer: Yes, haha, that's so much better, I completely forget the meaning of the sentence, just wanted it to sound good :see_no_evil:

Reviewer: I edited your comment's suggestion to fix a typo, pretty crazy workflow :scream:

> Overwrite the build process completely

Author: @ericholscher in a draft for a tweet in Front you mentioned that you would like to use something different than "Overwrite" for this. So, I'm pinging you here in case you have a better suggestion.

Reviewer 2: ```suggestion
Override the build process completely with build.commands
```

We call it override'ing in the docs, which is more correct: https://docs.readthedocs.io/en/stable/build-customization.html#override-the-build-process

======================================

> We are happy to announce a new *beta* feature that allow users to **overwrite the build process completely**.

Reviewer 2: ```suggestion
We are happy to announce a new *beta* feature that allow users to **override the Read the Docs build process**.
```

In the past, `we talked about executing custom commands <https://blog.readthedocs.com/user-defined-build-jobs/>`_ *in-between* the Read the Docs build process.

> However, that approach may not be sufficient for some projects with a heavily customized build process,

Reviewer 2: Is it just heavily customized builds? Or just any tool other than Sphinx or Mkdocs? We should likely mention both use cases.

Author: I added a mention to Pelican, Docsify and Docusaurus in the introduction.

> However, that approach may not be enough for some projects with strong custom build processes
> and they were not able to use our platform because of that.

Reviewer: ```suggestion
However, that approach isn't sufficient for projects with a heavily customized build processes,
some of which are not able to use our platform at all.
```

Author: I like your suggestion, but I would keep the "may" since there are projects with heavy customized builds where `build.jobs` is enough for them.

What about the following?

 ```suggestion
However, that approach may not be sufficient for some projects with a heavily customized build process,
some of which are not able to use our platform at all.
```

👍🏼 or 👎🏼 ?

Reviewer: :+1: or actually I just committed your suggestion :)

> In the past, `we talked about executing custom commands <https://blog.readthedocs.com/user-defined-build-jobs/>`_ *in-between* the Read the Docs build process.
> However, that approach may not be sufficient for some projects with a heavily customized build process,
> some of which are not able to use our platform at all.
> Until now! We have good news for them!

Reviewer 2: ```suggestion
We previously talked about `executing custom commands <https://blog.readthedocs.com/user-defined-build-jobs/>`_ *in-between* the Read the Docs build process.
That approach is not sufficient for projects with a heavily customized build process,
some of which are not able to use our platform at all.
Until now! We have good news for them!
```

> The new key ``build.commands`` allows projects to only execute *exactly* the commands they want.

Reviewer 2: ```suggestion
The new configuration file option ``build.commands`` allows projects to only execute *exactly* the commands they want.
```

`key` by itself doesn't mean anything to users.

No more. No less.

> This means that Read the Docs *won't execute any of the default* commands behind the scenes.

Reviewer 2: ```suggestion
This means that Read the Docs *won't execute any of the default commands* behind the scenes.
```

> You have 100% control over the building process.

Reviewer 2: ```suggestion
You have 100% control over the build process.
```

> How it works?
> -------------

Reviewer 2: ```suggestion
How does it work?
-------------------
```

To use the new *beta* feature,
you can specify all the commands the build process requires under the ``build.commands`` `config key <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-commands>`_.

> Finally, move the resulting documentation under ``_readthedocs/html`` directory.

Reviewer: ```suggestion
Finally, move the resulting documentation artifacts into the ``_readthedocs/html`` directory.
```

Read the Docs will upload and host all the content of that folder ✨

> Here is simple example that builds a small project's documentation using `Docsify <https://docsify.js.org/>`_:

Reviewer 2: I still feel like this isn't a great example given how different it is than Sphinx. It seems like we're kind of burying the lede here though. It seems like the announcement should be that other tools are now supported, not that you can override our build process completely?

Author: I didn't want to communicate the message "other tools are now supported" because they are not integrated yet. That message is going to be in future post after we have the integrations documented.

In this post I want to limit ourselves to communicate there is a new config option that allows people to override the build process. Leaving the door open to their imagination about what they can do with this.

We are still in an exploration stage and we are not sure what people want/is going to do with this.

We don't have _great_ examples at https://github.com/readthedocs/meta/discussions/28. Most of them are junky or can be replaced properly by `build.jobs`. Besides, the ones that we have documented in our official docs are Docsify and Pelican, so I think we should choose between one of them.

If the issue is not generating the HTML, we should go with the Pelican one.

Author: I changed the introduction a little to mention other doctools and use a Docusaurus example with a link to a live version of the resulting build. Docusaurus looks like a good example here since it has a completely different build process 👍🏼

Author: With this change, I think the post is better:

- limitation of the current fixed build process
- presenting a new config key to change that
- use cases for other doctools and projects with customized build processes
- explain how it works
- show a real example
- mention current limitations
- ask for feedback

💯 

Let me know what you think.

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-22.04"
     tools:
       python: "3.11"
     commands:
       # Create the output directory
       - mkdir --parents _readthedocs/html/
       # Copy the documents (including the ``index.html``)
       # into the output directory

>        - cp --recursive docs/* _readthedocs/html/

Reviewer: Would you agree that ideally `_readthedocs/html/` is read from some environment variable, like a `$READTHEDOCS_HTTP_ROOT` or similar? Just wanted to know if it would be relevant to open an issue, I don't see anything listed here: https://docs.readthedocs.io/en/latest/environment-variables.html

Author: I don't have strong opinions about this. This is the default contract that we built some time ago when working on this.

What would be the benefit about defining an environment variable? We could avoid copying a bunch of files; but instead of `cp` we could just create a symlink as well, if that's the problem we want to avoid. I'm not sure if an environment variable will be a benefit for users or will be trickier instead.

If you want to read more about ``build.commands``,
`read the full documentation <https://docs.readthedocs.io/en/latest/build-customization.html>`_.

Limitations
-----------

> As we mentioned before,
> this feature is still in *beta* and has some limitation currently.
> We are working towarsd making ``build.commands`` more integrated into our platform features.
> Mainly, the flyout with links to the Read the Docs project and all its active versions.
> We are close to get there,
> and we will publishing more content about this in the next months.
> Stay tunned!

Reviewer 2: ```suggestion
This feature is still in *beta* and has some limitation currently.
We are working to make ``build.commands`` more integrated into our platform features.
One major missing feature is the flyout which gives access to downloads, versions, and other functionality.
We are actively working on improving this functionality,
and we will publishing more content about this in the next months.
Stay tuned!
```

Try it out and give us feedback!
--------------------------------

We want to continue improving this *beta* feature.

> We think it has lot of potential and we want to build it in a way that's useful for our users.

Reviewer: ```suggestion
It has lot of potential and we want to develop it in a way that's useful for our users.
```

> .. post:: Nov 10, 2022

Author: ```suggestion
.. post:: Nov 11, 2022
```

> .. post:: Nov 11, 2022

Author: ```suggestion
.. post:: Nov 14, 2022
```

.. meta::
   :description lang=en:
      We released a new config key (``build.commands``) to specify user-defined commands
      which allows users to override the build process completely.

Override the build process completely with ``build.commands``
=============================================================

We are happy to announce a new *beta* feature that allows users to **override the Read the Docs build process completely**.
We previously talked about `executing custom commands <https://blog.readthedocs.com/user-defined-build-jobs/>`_ *in-between* the Read the Docs build process.

> That approach is not be sufficient for some projects with a heavily customized build processes,

Reviewer 2: ```suggestion
That approach is not sufficient for projects with a heavily customized build process,
```

> or those that want to use a different documentation tool like Pelican_, Docsify_ and Docusaurus_ for their documentation.

Reviewer 2: 💯

> Some of which are not able to use our platform at all.

Reviewer 2: ```suggestion
Some of which were not able to use our platform at all.
```

They can now!

No more. No less.
This means that Read the Docs *won't execute any of the default commands* behind the scenes.
You have 100% control over the build process.

How does it work?
-----------------

To use the new *beta* feature,
you can specify all the commands the build process requires under the ``build.commands`` `config key <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-commands>`_.
Finally, move the resulting documentation artifacts into the ``_readthedocs/html`` directory.
Read the Docs will upload and host all the content of that folder ✨

Here is simple example that builds the Docusaurus tutorial:

.. code:: yaml

   version: 2

   build:
     os: ubuntu-22.04
     tools:
       nodejs: "18"
     commands:
       # "docs/" directory was created using the command to create the site:
       # npx create-docusaurus@latest docs classic
       #
       # Install Docusaurus dependencies
       - cd docs/ && npm install
       # Build the site
       - cd docs/ && npm run build
       # Copy generated files into Read the Docs directory
       - mkdir --parents _readthedocs/html/
       - cp --recursive docs/build/* _readthedocs/html/

> The resulting documentation lives at https://test-builds.readthedocs.io/en/docusaurus/.

Reviewer 2: This is great too! 💯
